extensions/wikia/IndexingPipeline/PipelineEventProducer.class.php

```diff
@@ -2,68 +2,201 @@
+
 		return true;
 	}
 
-	static public function onArticleDeleteComplete( &$oPage, &$oUser, $reason, $pageId ) {
-		wfDebug( "IndexingPipeline:onArticleDeleteComplete\n" );
+	/**
+	 * @desc Fires on:
+	 *  - article delete
+	 * @return bool
+	 */
+	public static function onArticleDeleteComplete( &$oPage, &$oUser, $reason, $pageId ) {
+		$ns = self::preparePageNamespaceName( $oPage->getTitle() );
+
 		self::send( 'onArticleDeleteComplete', $pageId );
+		self::nSend( self::ACTION_DELETE, $pageId, $ns );
+
 		return true;
 	}
 
-	static public function onArticleUndelete( Title &$oTitle, $isNew = false ) {
-		wfDebug( "IndexingPipeline:onArticleUndelete\n" );
-		self::send( 'onArticleUndelete', $oTitle->getArticleId(), [ 'isNew' => $isNew ] );
+	/**
+	 * @desc Fires on:
+	 *  - article restore
+	 * Send ACTION_CREATE as an article with new ID is created
+	 * @return bool
+	 */
+	public static function onArticleUndelete( Title &$oTitle, $isNew = false ) {
+		$ns = self::preparePageNamespaceName( $oTitle );
+		$data = [ 'isNew' => $isNew ];
+
+		self::send( 'onArticleUndelete', $oTitle->getArticleId(), $data );
+		self::nSend( self::ACTION_CREATE, $oTitle->getArticleId(), $ns, $data );
+
 		return true;
 	}
 
-	static public function onTitleMoveComplete( &$oOldTitle, &$oNewTitle, &$oUser, $pageId, $redirectId = 0 ) {
-		wfDebug( "IndexingPipeline:onTitleMoveComplete\n" );
-		self::send( 'onTitleMoveComplete', $pageId, [ 'redirectId' => $redirectId ] );
+	/**
+	 * @desc Fires on:
+	 *  - article rename
+	 * Send ACTION_UPDATE as the ID of article remains the same only title changes
+	 * @return bool
+	 */
+	public static function onTitleMoveComplete( &$oOldTitle, &$oNewTitle, &$oUser, $pageId, $redirectId = 0 ) {
+		$ns = self::preparePageNamespaceName( $oNewTitle );
+		$data = [ 'redirectId' => $redirectId ];
+
+		self::send( 'onTitleMoveComplete', $pageId, $data );
+		self::nSend( self::ACTION_UPDATE, $pageId, $ns, $data );
+
 		return true;
 	}
+
+	/*
+	 * Helper methods
+	 */
+
+	/**
+	 * @desc create message with cityId and pageId fields
+	 * @param $pageId
+	 * @return \stdClass
+	 */
+	protected static function prepareMessage( $pageId ) {
+		global $wgCityId;
+		$msg = new stdClass();
+		$msg->cityId = $wgCityId;
+		$msg->pageId = $pageId;
+
+		return $msg;
+	}
+
+	/**
+	 * @desc create message route
+	 * @param $action
+	 * @param $ns
+	 * @param $flags
+	 * @param $data
+	 * @return string message route in format:
+	 * PRODUCER_NAME.ROUTE_ACTION_KEY:{action}.ROUTE_NAMESPACE_KEY:{namespace}.ROUTE_CONTENT_KEY:{items}
+	 */
+	public static function prepareRoute( $action, $ns, $flags, $data ) {
+		//prepare info about the data is in the message content
+		$contentData = array_map( function ( $item ) {
+				return self::ROUTE_CONTENT_KEY . ':' . $item;
+			}, array_keys( $data ) );
+
+		$routeData = array_merge(
+			[ self::PRODUCER_NAME, self::ROUTE_ACTION_KEY . ':' . $action,  self::ROUTE_NAMESPACE_KEY . ':' . $ns ],
+			$flags,
```

Comment:
Per the remark in the tests, you may need to introduce some more checks / ensure the params are of the right type